src/sys/fuchsia/eventedfd.rs

```diff
@@ -0,0 +1,257 @@
+use {io, poll, Evented, Ready, Poll, PollOpt, Token};
+use libc;
+use magenta;
+use magenta::HandleBase;
+use sys::fuchsia::{DontDrop, poll_opts_to_wait_async, sys};
+use std::mem;
+use std::os::unix::io::RawFd;
+use std::sync::{Arc, Mutex};
+
+/// Properties of an `EventedFd`'s current registration
+#[derive(Debug)]
+pub struct EventedFdRegistration {
+    token: Token,
+    handle: DontDrop<magenta::Handle>,
+    rereg_signals: Option<(magenta::Signals, magenta::WaitAsyncOpts)>,
+}
+
+impl EventedFdRegistration {
+    unsafe fn new(token: Token,
+                  raw_handle: sys::mx_handle_t,
+                  opts: PollOpt,
+                  signals: magenta::Signals) -> Self
+    {
+        let needs_rereg = opts.is_level() && !opts.is_oneshot();
+
+        EventedFdRegistration {
+            token: token,
+            handle: DontDrop::new(magenta::Handle::from_raw(raw_handle)),
+            rereg_signals: if needs_rereg {
+                Some((signals, poll_opts_to_wait_async(opts)))
+            } else {
+                None
+            },
+        }
+    }
+
+    pub fn rereg_signals(&self) -> Option<(magenta::Signals, magenta::WaitAsyncOpts)> {
+        self.rereg_signals
+    }
+}
+
+/// An event-ed file descriptor. The file descriptor is owned by this structure.
+#[derive(Debug)]
+pub struct EventedFdInner {
+    /// Properties of the current registration.
+    registration: Mutex<Option<EventedFdRegistration>>,
+
+    /// Owned file descriptor.
+    ///
+    /// `fd` is closed on `Drop`, so modifying `fd` is a memory-unsafe operation.
+    fd: RawFd,
+
+    /// Owned `mxio_t` ponter.
+    mxio: *const sys::mxio_t,
+}
+
+impl EventedFdInner {
+    pub fn rereg_for_level(&self, port: &magenta::Port) {
+        let registration_opt = self.registration.lock().unwrap();
+        if let Some(ref registration) = *registration_opt {
+            if let Some((rereg_signals, rereg_opts)) = registration.rereg_signals {
+                let _res =
+                    registration
+                        .handle.inner_ref()
+                        .wait_async(port,
```

Comment:
This concern is now moot since @cramertj switched to using one shots after I wrote this.